Add request body minimum data rate feature

cesarbs opened the pull request:
TODO:

- [x] Deal with backpressure
- [x] Add tests for backpressure
- [x] Better logs
- [x] Tests
- [x] Validate timeout values on server init
- [x] Add feature to make timeout configurable per request
- [x] Only start timing reads after sending 100 Continue.
- [x] Don't enforce timeout on upgrade requests

davidfowl commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs:
```diff
@@ -231,12 +235,34 @@ private void DisposeAdaptedConnections()
-            if (timestamp > Interlocked.Read(ref _timeoutTimestamp))
+            if (_meteringReads)
+            {
+                var elapsed = TimeSpan.FromTicks(timestamp - _meteringStartTicks);
```
Why is this logic in FrameConnection and not frame?

natemcmaster commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/RequestBodyTimeout.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core
+{
+    public class RequestBodyTimeout
```
Needs xml docs. The API names alone don't provide enough explanation about what these mean and how to set them.

Also, we should provide some validation of inputs, such as MaxTime must be > MinTime, MinimumRate must be >= 0, etc. Consider making the properties readonly and adding a constructor that does these validations.

cesarbs replied:
Yep, I'll validate that in `KestrelServer` like we do with other limits.

natemcmaster commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -207,5 +209,7 @@ public TimeSpan RequestHeadersTimeout
             }
         }
+
+        public RequestBodyTimeout DefaultRequestBodyTimeout => _defaultRequestBodyTimeout;
```
xml docs here too

cesarbs commented:
@Tratcher The extended timeout and minimum rate must be set at the same time. And I didn't want to allow for inconsistensies (e.g. extended timeout shorter than the "base" timeout).

cesarbs commented:
@Tratcher This is implementing the same as https://httpd.apache.org/docs/2.4/mod/mod_reqtimeout.html#requestreadtimeout.

cesarbs commented:
@Tratcher Commit 17ca98882a3e6cdc19171ab6b6ad22ef4929d064 disables the timeout on upgrade requests.

cesarbs commented:
Latest commit (5e90345b6e8cb0e406212854432a499e3a13d2ab) changes the behavior so far to send a 408 instead of aborting the connection.

Tratcher commented:
How does that help? Will you then try to drain the timed out request body at the end of the request processing?

Tratcher commented:
You still have to close the connection.

cesarbs commented:
@Tratcher It will either throw in the app (on a read) or when attempting to consume the rest of the request body (if the app catches the exception). Either way the connection will be closed.

natemcmaster commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServer.cs:
```diff
@@ -208,6 +208,19 @@ private void ValidateOptions()
+                    CoreStrings.FormatRequestBodyExtendedTimeoutSmallerThanTimeout(Options.Limits.DefaultRequestBodyExtendedTimeout, Options.Limits.DefaultRequestBodyTimeout));
+            }
+
+            if (Options.Limits.DefaultRequestBodyExtendedTimeout.HasValue ^ Options.Limits.DefaultRequestBodyMinimumDataRate.HasValue)
```
Since they are both required to be null or set, can we make this enforceable as a compile-time check? Makes the API slightly easier to use

Ideas:
```c#
// add a type
public class KestrelServerLimits
{
    public DataRateLimit DefaultRequestBodyRate { get; set; }
}

public class DataRateLimit
{
    public TimeSpan Timeout { get;set;}
    public double Rate { get; set; }
}

```
```c#
// add a tuple literal
public class KestrelServerLimits
{
    public (TimeSpan timespan, double rate) DefaultRequestBodyRate { get; set; }
}
```
```c#
// add a tuple type
public class KestrelServerLimits
{
    public Tuple<TimeSpan, double> DefaultRequestBodyRate { get; set; }
}
```

Tratcher commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Features/IHttpRequestBodyTimeoutFeature.cs:
```diff
@@ -0,0 +1,46 @@
+        /// The maximum amount of time allowed for the request body to be entirely read.
+        /// Only effective when <see cref="MinimumDataRate"/> is also set.
+        /// </summary>
+        TimeSpan? ExtendedTimeout { get; }
```
Still haven't resolved the redundancy between ExtendedTimeout and MinimumDataRate. MinimumDataRate scales with the size of the body, ExtendedTimeout doesn't. ExtendedTimeout will be a pain for any customer that tries to use it.

cesarbs commented:
Made changes to the API as discussed in person. Now we have:

- Request body timeout: max. amount of time allowed for the request body to be received
- Request body minimum data rate: self-explanatory
- Request body minimum data rate grace period: the initial period of time during which the minimum rate is not enforced

There was no need for a new data structure. I added the `SetDefaultRequestBodyMinimumDataRate(double, TimeSpan)` method to `KestrelServerLimits` to configure the default minimum rate, and changed `IHttpRequestBodyTimeoutFeature` to have the analogous `SetMinimumDataRate(double, TimeSpan)` method. The overall request body timeout is not a settable property in the feature.

natemcmaster commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Features/IHttpRequestBodyTimeoutFeature.cs:
```diff
@@ -0,0 +1,35 @@
+        TimeSpan RequestBodyTimeout { get; set; }
+
+        /// <summary>
+        /// The minimum data rate at which the request body should be received.
```
Would be good to specify the units. bytes/second? MB/min? etc.

natemcmaster commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Features/IHttpRequestBodyTimeoutFeature.cs:
```diff
@@ -0,0 +1,35 @@
+        double MinimumDataRate { get; }
+
+        /// <summary>
+        /// The period of time during which the minimum data rate will not be enforced.
```
Suggested change of wording: `The amount of time to delay enforcement of <see cref="MinimumDataRate" />. When set to <see cref="TimeSpan.Zero"/>, enforcement begins when the server starts reading the request body.`

natemcmaster commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -207,5 +212,50 @@ public TimeSpan RequestHeadersTimeout
+        /// <summary>
+        /// Sets the default request body minimum data rate. No minimum data rate is enforced by default.
+        /// </summary>
+        public void SetDefaultRequestBodyMinimumDataRate(double minimumDataRate, TimeSpan gracePeriod)
```
Add `<param>` elements to the docs

davidfowl commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/Frame.cs:
```diff
@@ -298,6 +298,12 @@ private CancellationTokenSource RequestAbortedSource
+
+        public TimeSpan RequestBodyMinimumDataRateGracePeriod { get; set; }
+
         public void InitializeStreams(MessageBody messageBody)
```
nit: Probably should renamed to `InitializeBody`

cesarbs commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs:
```diff
@@ -215,6 +245,26 @@ private void TryProduceContinue()
+
+        private void TryStopTimingReads()
+        {
+            if (!RequestUpgrade)
```
That bool is not necessary and I'm removing it.

cesarbs commented:
### Benchmarks

#### Plaintext

rel/2.0.0-preview2: 1201645.23
PR: 1194168.72

#### CopyToAsync

rel/2.0.0-preview2: 10706.16
PR: 10669.96

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs:
```diff
@@ -231,21 +255,49 @@ private void DisposeAdaptedConnections()
-                _frame.Stop();
+                // PauseTimingReads() cannot just set _timingReads to false. It needs to go through at least one tick
+                // before pausing, otherwise _readTimingElapsed might never be updated if PauseTimingReads() is always
+                // called before the next tick.
```
This means that elapsed time is being added while there is no read callback configured. This means about a second of elapsed time could be added for each pause when the client actually can't upload additional data.

If we error, we should error the other way.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs:
```diff
@@ -275,5 +327,35 @@ private void AssignTimeout(long ticks, TimeoutAction timeoutAction)
+        {
+            Interlocked.Exchange(ref _readTimingElapsed, 0);
+            Interlocked.Exchange(ref _bytesRead, 0);
+            Interlocked.Exchange(ref _readTimingStartTicks, _lastTimestamp);
```
Why are these Interlocked.Exchange's helping? If we're worried about concurrent access/modification of these fields, wouldn't these fields being momentarily inconsistent with each other be a problem?

You can't safely do `+=` operations to fields that are reset concurrently even with Interlocked.Exchange, so `_readTimingElapsed += timestamp - _lastTimestamp;` seems unsafe. I don't think setting `_timingReads` to true last saves you here.

cesarbs replied:
Yeah these are unnecessary. Setting `_timingReads` last should help since the other properties won't be read until that is true.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs:
```diff
@@ -275,5 +327,35 @@ private void AssignTimeout(long ticks, TimeoutAction timeoutAction)
+        public void PauseTimingReads()
+        {
+            Log.RequestBodyTimingPause(_context.ConnectionId, _frame.TraceIdentifier, TimeSpan.FromTicks(_readTimingElapsed), TimeSpan.FromTicks(_lastTimestamp - _readTimingStartTicks));
+            _pauseTimingReads = true;
```
I'm pretty sure you're going to need to add a lock to FrameConnection. The locks would be per-connection and should always be for quick operations, but I think we could use Monitor.TryEnter in Tick if need be.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs:
```diff
@@ -231,21 +255,49 @@ private void DisposeAdaptedConnections()
                 }
+                else if (_frame.RequestBodyMinimumDataRate > 0)
+                {
+                    var rate = (_bytesRead / (_readTimingElapsed / TimeSpan.TicksPerSecond));
```
I would clarify this line of code by splitting it into a couple lines, ex:

```
var elapsedSecondsFloored = _readTimingElapsed / TimeSpan.TicksPerSecond;
var rate = _bytesRead / elapsedSecondsFloored ;
```

Technically the rate is floored too which isn't exactly ideal, but bytes/second is a precise enough measurement that I doubt it matters.

cesarbs replied:
Doesn't need to be floored really, I'll cast to `double`.

Tratcher commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Features/IHttpRequestBodyTimeoutFeature.cs:
```diff
@@ -0,0 +1,39 @@
+        /// <summary>
+        /// The maximum amount of time in which the request body should be fully received.
+        /// </summary>
+        TimeSpan RequestBodyTimeout { get; set; }
```
Are you mirroring apache's incremental timeout behavior?
"If the client sends data, increase the timeout by 1 second for every 1000 bytes received, with no upper limit for the timeout" - https://httpd.apache.org/docs/2.4/mod/mod_reqtimeout.html
Otherwise this feature is going to be very difficult to use.

cesarbs replied:
Yes. We check the rate every second and time out if it drops below the minimum required, so that's effectively the same thing mod_reqtimeout does.

Tratcher replied:
That doesn't sound the same at all. apache's full body timeout gets longer so long as data is still being received.

cesarbs replied:
Up to a limit. You give it a range (e.g. 20-40 seconds) and it will increase the timeout by 1 second up to the upper bound as long as data is arriving at the specified rate.

Tratcher commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs:
```diff
@@ -242,10 +267,40 @@ public void Tick(DateTimeOffset now)
+            {
+                _readTimingElapsed += timestamp - _lastTimestamp;
+
+                if (_readTimingElapsed > _frame.RequestBodyTimeout.Ticks)
```
Where do you grow RequestBodyTimeout like Apache's?

halter73 replied:
My understanding is that RequestBodyTimeout is unaffected by rate. This is the absolute max timeout. If you want a timeout that grows with more data, you could just set a rate and leave this unlimited.

Tratcher replied:
Ah, those docs are just talking about the rate.

I don't see how an absolute timeout is workable for variable payload sizes. You'd always have to change the size limit and the time limit in tandem, and the time limit is only a multiple of the data rate and the size limit. What value does RequestBodyTimeout add beyond the size and rate limits?

halter73 replied:
This along with RequestHeadersTimeout gives the developer a simple/convenient way to set an absolute timeout for a client to send a complete request. Even if it's possible to achieve the same thing by doing some math and selecting the appropriate min rate and max size, that seems more complicated than necessary for many scenarios.

Tratcher replied:
You have to do the math anyways don't you? When you set this you have to make sure you allow enough time for your expected data size to arrive at the expected rate. RequestBodyTimeout is completely redundant.

halter73 replied:
You don't have to do the math. You could just limit one of rate/timeout/size and set the others unlimited if that's want you want.

Tratcher replied:
I'll accept as much data as you can send me in 60s? I don't see that as a realistic scenario.

A full request/response processing time limit seems more realistic. E.g. my SLA for this request is 1s, and if I can't fulfill it I need to cancel.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -207,5 +210,26 @@ public TimeSpan RequestHeadersTimeout
+            get => _defaultRequestBodyTimeout;
+            set
+            {
+                if (value <= TimeSpan.Zero)
```
Shouldn't we support `Timeout.InfiniteTimeSpan`? `Timeout.InfiniteTimeSpan <= TimeSpan.Zero` BTW.

halter73 replied:
Add this check to KeepAliveTimeout and RequestHeadersTimeout while you're at it.

Tratcher replied:
Timeout.InfiniteTimeSpan (-1s) to disable?

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -207,5 +210,26 @@ public TimeSpan RequestHeadersTimeout
+
+        /// <summary>
+        /// Gets or sets the default timeout period for receiving the request body. Defaults to 2 minutes.
+        /// </summary>
```
Nit: Default values are in the remarks section of the other properties.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -207,5 +210,26 @@ public TimeSpan RequestHeadersTimeout
+        /// <summary>
+        /// Gets or sets the default timeout period for receiving the request body. Defaults to 2 minutes.
+        /// </summary>
+        public TimeSpan DefaultRequestBodyTimeout
```
Rename to just RequestBodyTimeout since it can be modified to a non-default value. I get that this value can be modified on a per-request basis, but that can be called out in the doc comments [like this](https://github.com/aspnet/KestrelHttpServer/blob/3dbfb003bb064e117b52d6336ee145cf602363ea/src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs#L145). This is better because it explains how to change the "default".

Tratcher replied:
Timeout.InfiniateTimeSpan (-1s) to disable?

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -207,5 +210,26 @@ public TimeSpan RequestHeadersTimeout
+        /// <summary>
+        /// Gets or sets the default request body minimum data rate in bytes/second.
+        /// </summary>
+        public MinimumDataRate DefaultRequestBodyMinimumDataRate { get; set; } = null;
```
Also remove "Default" from this property name.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -207,5 +210,26 @@ public TimeSpan RequestHeadersTimeout
+
+        /// <summary>
+        /// Gets or sets the default request body minimum data rate in bytes/second.
+        /// </summary>
```
The summary should mention that null means there is no minimum rate, and add a remark indicating that null is the default value.

Tratcher replied:
The spec calls for a non-null default.

cesarbs replied:
What spec?

Tratcher replied:
https://github.com/aspnet/Coherence-Signed/issues/427 calls for a default read timeout. You've replaced that with a data rate, but the rate needs to be enabled by default still.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs:
```diff
@@ -242,10 +268,46 @@ public void Tick(DateTimeOffset now)
+                        // called before the next tick.
+                        if (_readTimingPauseRequested)
+                        {
+                            _readTimingEnabled = false;
```
What happens if both PauseTimingReads and ResumeTimingReads are called between a single tick?.

cesarbs replied:
fccfa37d24e52a9aede689f0b6bcf208f50f403b

cesarbs commented:
Updated to remove request body timeout, leaving only the min rate stuff in there.

Tratcher commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -234,5 +250,16 @@ public TimeSpan RequestHeadersTimeout
+        /// <remarks>
+        /// Defaults to null.
+        /// </remarks>
+        public MinimumDataRate RequestBodyMinimumDataRate { get; set; } = new MinimumDataRate(1, TimeSpan.FromSeconds(5));
```
use named parameters when passing in constants

Tratcher commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/MinimumDataRate.cs:
```diff
@@ -0,0 +1,41 @@
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core
+{
+    public class MinimumDataRate
```
Put this in the feature namespace?

Tratcher commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Features/IHttpRequestBodyMinimumDataRateFeature.cs:
```diff
@@ -0,0 +1,18 @@
+        /// Setting this property to null indicates no minimum data rate should be enforced.
+        /// This limit has no effect on upgraded connections which are always unlimited.
+        /// </summary>
+        MinimumDataRate MinimumDataRate { get; set; }
```
Design note: Using local concrete data types on a feature abstraction makes it hard to generalize / replace. This is tolerable so long as this remains a kestrel exclusive feature, but we'd need a different design if this were ever moved to HttpAbstractions.

cesarbs replied:
Couldn't we move `MinimumDataRate` to HttpAbstractions if that were the case? But yeah this is what I had in mind when I was opposing the new data type.

Tratcher replied:
Yes, the type could move with it, but it's existance still limits the extensibility of the feature. I think we only have one other local concrete type exposed from a feature, CookieOptions.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/KestrelServerLimits.cs:
```diff
@@ -169,7 +171,14 @@ public int MaxRequestHeaderCount
-            set => _keepAliveTimeout = value;
+            set
+            {
+                if (value <= TimeSpan.Zero && value != Timeout.InfiniteTimeSpan)
```
InfiniteTimeSpan needs to be special cased in the timeout logic. I'd add a functional test that verifies that connections aren't killed immediate when InfiniteTimeSpan is configured.

cesarbs replied:
I'm checking for it and setting to `TimeSpan.MaxValue`. Added unit tests.

halter73 commented on src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/FrameConnection.cs:
```diff
@@ -29,13 +29,25 @@ public class FrameConnection : IConnectionContext, ITimeoutControl
+        private object _readTimingLock = new object();
+        private bool _readTimingEnabled;
+        private bool _readTimingPauseRequested;
+        private long _readTimingStartTicks;
```
No longer needed.
